tests/src/deviceLib/hipAsynchronousStreams.cpp

```diff
@@ -0,0 +1,607 @@
+
+        if (flags & __OCKL_AS_CONNECTION_END) {
+            strRecd.insert(buf);
+            buffers.erase(connection_id);
+        }
+    }
+
+    if (strRecd.size() != 4) return false;
+    if (strRecd.count(std::string(STR30)) == 0) return false;
+    if (strRecd.count(std::string(STR47)) == 0) return false;
+    if (strRecd.count(std::string(STR95)) == 0) return false;
+    if (strRecd.count(std::string(STR124)) == 0) return false;
+
+    return true;
+}
+
+bool ocklAsMixedProducers() {
+    bool success = true;
+    unsigned int numThreads = 4;
+    unsigned int numBlocks = 1;
+
+    unsigned int numPackets = 1024;
+    unsigned int bufferSize = sizeof(__ockl_as_stream_t) + numPackets * __OCKL_AS_PACKET_SIZE;
+
+    void* buffer;
+    HIPCHECK(hipHostMalloc(&buffer, bufferSize));
+
+    __ockl_as_signal_t doorbell_signal = {0};
+    __ockl_as_stream_t* stream =
+        __ockl_as_stream_create(buffer, bufferSize, numPackets, doorbell_signal);
+
+    hipLaunchKernelGGL(mixedProducers, dim3(numBlocks), dim3(numThreads), 0, 0, stream);
+
+    HIPCHECK(hipDeviceSynchronize());
+
+    if (!checkMixedProducers(stream)) {
+        test_failed(__func__);
+        success = false;
+    }
+
+    HIPCHECK(hipHostFree(buffer));
+    return success;
+}
+
+#define STR27 "In et consectetur mi metus."
+#define STR64 "Praesent tempus arcu id ligula blandit, eget congue justo metus."
+#define STR40 "Sed at dolor ipsum. Curabitur cras amet."
+
+__global__ void splitMessage(__ockl_as_stream_t* stream) {
+    const char* str27 = STR27;
+    const char* str64 = STR64;
+    const char* str40 = STR40;
+
+    uint64_t connection_id;
+    __ockl_as_write_block(stream, 42, &connection_id, (unsigned const char*)str27, 27,
+                          __OCKL_AS_CONNECTION_BEGIN);
+    __ockl_as_write_block(stream, 42, &connection_id, (unsigned const char*)str64, 64, 0);
+    __ockl_as_write_block(stream, 42, &connection_id, (unsigned const char*)str40, 40,
+                          __OCKL_AS_CONNECTION_END);
+}
+
+bool checkSplitMessage(__ockl_as_stream_t* stream) {
+    std::string buffer;
+
+    for (unsigned long read_index = 0; read_index != stream->write_index; ++read_index) {
+        __ockl_as_packet_t* packet = stream->base_address + read_index;
+
+        uint header = packet->header;
+        if (get_packet_type(header) != __OCKL_AS_PACKET_READY) return false;
+
+        uint bytes = get_packet_bytes(header);
+        unsigned char flags = get_packet_flags(header);
+        unsigned long connection_id = packet->connection_id;
+        unsigned char* payload = packet->payload;
+
+        if ((flags & __OCKL_AS_CONNECTION_BEGIN) != (buffer.size() == 0)) return false;
+
+        buffer.insert(buffer.end(), payload, payload + bytes);
+
+        if (flags & __OCKL_AS_CONNECTION_END) break;
+    }
+
+    if (buffer.compare(STR27 STR64 STR40) != 0) return false;
+
+    return true;
+}
+
+bool ocklAsSplitMessage() {
```

Comment:
Need a multiple split messages test with a non-tiny number of threads, and blocks.